src/ir/auth_logic/universe_relation_insertion.cc

```diff
@@ -0,0 +1,357 @@
+// identifiers are a separate AST node with constant and non-constant
+// children determined at parse time rather than here since
+// this information is available at parse time.
+bool IsNameConstant(absl::string_view id) {
+  if (id[0] == '"') {
+    return true;
+  } else {
+    // Numeric literals are constants and this is
+    // a numeric literal if all the characters are numbers.
+    for (char c : id) {
+      if (!isdigit(c)) return false;
+    }
+    return true;
+  }
+}
+
+// (TODO: #672) once an AST node for numbers as RVALUEs in numeric
+// comparisons is added, this visitor should also visit
+// the numeric RVALUEs and add them to the type environment
+// with type ArgumentType(kNumber, "Number")
+// This is a workaround that makes use of the fact that
+// numeric comparisons are represented as predicates
+// with a name that matches the operator:
+// (https://github.com/google-research/raksha/blob/be6ef8e1e1a20735a06637c12db9ed0b87e3d2a2/src/ir/auth_logic/ast_construction.cc#L92)
+static inline bool PredicateIsNumericOperator(const datalog::Predicate& pred) {
+  if (pred.name() == "<" || pred.name() == "<" || pred.name() == "=" ||
+      pred.name() == "!=" || pred.name() == "<=" || pred.name() == ">=") {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+std::string TypeToUniverseName(datalog::ArgumentType arg_type) {
+  switch (arg_type.kind()) {
+    case datalog::ArgumentType::Kind::kNumber:
+      return "isNumber";
+    case datalog::ArgumentType::Kind::kPrincipal:
+      return "isPrincipal";
+    case datalog::ArgumentType::Kind::kCustom:
+      return absl::StrCat("is", arg_type.name());
+  }
+}
+
+BaseFact MakeUniverseMembershipFact(std::string element_name,
+                                    datalog::ArgumentType arg_type) {
+  return BaseFact(
+      datalog::Predicate(TypeToUniverseName(arg_type),  // universe relation
+                         {element_name},  // the element to add to universe
+                         datalog::Sign::kPositive));
+}
+
+datalog::RelationDeclaration TypeToUniverseDeclaration(
+    datalog::ArgumentType arg_type) {
+  return datalog::RelationDeclaration(TypeToUniverseName(arg_type), false,
+                                      {datalog::Argument("x", arg_type)});
+}
+
+// This generates a set of says assertions that populate all the
+// universes for all the speakers given the type environment. Essentially,
+// it loops through the list of constant principals, and then all the
+// constants of all types, and then generates a SaysAssertion that
+// adds the constant to the universe for each speaker.
+std::vector<SaysAssertion> GetUniverseDefiningFacts(
+    const TypeEnvironment& type_env) {
+  std::vector<Principal> constant_principals;
+  for (const auto& typing : type_env.inner_map()) {
+    if (typing.second.kind() == datalog::ArgumentType::Kind::kPrincipal) {
+      constant_principals.push_back(Principal(typing.first));
+    }
+  }
```

Comment:
Nothing for you to do on this comment, but seeing this code makes me want to put a functional-style `filter` in our `//src/common/utils`.